tests/unit/CollectionGenericOperationTest.php

```diff
@@ -0,0 +1,173 @@
+     * @dataProvider prependOperationProvider
+     * @dataProvider productOperationProvider
+     * @dataProvider reduceOperationProvider
+     * @dataProvider reductionOperationProvider
+     * @dataProvider rejectOperationProvider
+     * @dataProvider reverseOperationProvider
+     * @dataProvider scaleOperationProvider
+     * @dataProvider scanLeftOperationProvider
+     * @dataProvider scanLeft1OperationProvider
+     * @dataProvider scanRightOperationProvider
+     * @dataProvider scanRight1OperationProvider
+     * @dataProvider shuffleOperationProvider
+     * @dataProvider sinceOperationProvider
+     * @dataProvider sliceOperationProvider
+     * @dataProvider sortOperationProvider
+     * @dataProvider splitOperationProvider
+     * @dataProvider squashOperationProvider
+     * @dataProvider strictOperationProvider
+     * @dataProvider tailOperationProvider
+     * @dataProvider tailsOperationProvider
+     * @dataProvider takeWhileOperationProvider
+     * @dataProvider transposeOperationProvider
+     * @dataProvider unlinesOperationProvider
+     * @dataProvider unpackOperationProvider
+     * @dataProvider unpairOperationProvider
+     * @dataProvider untilOperationProvider
+     * @dataProvider unwindowOperationProvider
+     * @dataProvider unwordsOperationProvider
+     * @dataProvider unwrapOperationProvider
+     * @dataProvider unzipOperationProvider
+     * @dataProvider whenOperationProvider
+     * @dataProvider windowOperationProvider
+     * @dataProvider wordsOperationProvider
+     * @dataProvider wrapOperationProvider
+     * @dataProvider zipOperationProvider
+     */
+    public function testGeneric(
+        string $operation,
+        array $parameters,
+        iterable $actual,
+        iterable $expected,
+        int $limit = 0
+    ): void {
+        $actual = Collection::fromIterable($actual)->{$operation}(...$parameters);
+
+        if (0 !== $limit) {
+            $actual = $actual->limit($limit);
+        }
+
+        $this::assertIdenticalIterable(
+            $expected,
+            $actual
+        );
+    }
+
+    /**
+     * @dataProvider containsOperationProvider
+     * @dataProvider countOperationProvider
+     * @dataProvider currentOperationProvider
+     * @dataProvider equalsOperationProvider
+     * @dataProvider everyOperationProvider
+     * @dataProvider falsyOperationProvider
+     * @dataProvider findOperationProvider
+     * @dataProvider hasOperationProvider
+     * @dataProvider isEmptyOperationProvider
+     * @dataProvider keyOperationProvider
+     * @dataProvider matchOperationProvider
+     * @dataProvider nullsyOperationProvider
+     * @dataProvider sameOperationProvider
+     * @dataProvider truthyOperationProvider
+     *
+     * @param mixed $expected
+     */
+    public function testGenericScalar(
+        string $operation,
+        array $parameters,
+        iterable $actual,
+        $expected
+    ): void {
+        self::assertEquals(
```

Comment:
```suggestion
        self::assertSame(
```

I would recommend we avoid `assertEquals` altogether, and use `assertSame` for scalars because it's more akin to `===`, checking both the value and the type. For objects we should use [assertObjectEquals](https://phpunit.readthedocs.io/en/9.5/assertions.html#assertobjectequals)